src/core/regimes.rkt

```diff
@@ -175,127 +175,131 @@
      (and (> pidx 0)) (list-ref can-split? pidx))
    (= (si-pidx (last split-indices)) (length can-split?))))
 
+;; This is the core main loop of the regimes algorithm.
+;; Takes in a list of alts in the form of there error at a given point
+;; as well as a list of split indices to determine when it's ok to split
+;; for another alt.
+;; Returns a list of split indices saying which alt to use for which
+;; range of points. Starting at 1 going up to num-points.
+;; Alts are indexed 0 and points are index 1.
 (define/contract (err-lsts->split-indices err-lsts can-split)
   (->i ([e (listof list)] [cs (listof boolean?)]) 
         [result (cs) (curry valid-splitindices? cs)])
+  ;; Coverts the list to vector form for faster processing
   (define can-split-vec (list->vector can-split))
-  (define err-lsts-vec (list->vector err-lsts))
-  ;; We have num-candidates candidates, each of whom has error lists of length num-points.
-  ;; We keep track of the partial sums of the error lists so that we can easily find the cost of regions.
-  (define num-candidates (vector-length err-lsts-vec))
-  (define num-points (vector-length can-split-vec))
-  (define min-weight (fl num-points))
-
+  ;; Converting list of list to list of flvectors
+  ;; flvectors are used to remove pointer chasing
   (define (make-vec-psum lst) 
    (flvector-sums (list->flvector lst)))
-  (define flvec-psums (vector-map make-vec-psum err-lsts-vec))
+  (define flvec-psums (vector-map make-vec-psum (list->vector err-lsts)))
+
+  ;; Set up data needed for algorithm
+  (define number-of-alts (vector-length flvec-psums))
+  (define number-of-points (vector-length can-split-vec))
+  ;; min-weight is used as penalty to favor not adding split points
+  (define min-weight (fl number-of-points))
+  
+  ;; These 3 vectors are will contain the output data and be used for
+  ;; determining which alt is best for a given point
+  (define result-error-sums (make-flvector number-of-points))
+  (define result-alt-idxs (make-vector number-of-points))
+  (define result-prev-idxs (make-vector number-of-points))
+
+  ;; Temporary vector used to find the alt with least error for each point
+  (define vec-temp (make-flvector number-of-alts))
+  ;; TODO invert loops to match core algorithm data priority
+  (for ([point-idx (in-range number-of-points)])
+    ;; record the error for each candidate
+    (for ([alt-idx (range number-of-alts)])
+     (flvector-set! vec-temp alt-idx
+      (flvector-ref (vector-ref flvec-psums alt-idx) point-idx)))
+    ;; Find the min, no built in function to find smallest fl in vector
+    (define min-v (flvector-ref vec-temp 0))
+    (define min-idx 0)
+    (for ([val vec-temp] [idx (range number-of-alts)])
+      (cond [(< val min-v)
+             (set! min-idx idx)
+             (set! min-v val)]))
+    (flvector-set! result-error-sums point-idx (fl min-v))
+    (vector-set! result-alt-idxs point-idx min-idx)
+    (vector-set! result-prev-idxs point-idx number-of-points))
```

Comment:
Right now you're computing a min into variables and then copying the variables into a vector. Maybe that's faster but it might be faster (and shorter) to just compute the min directly into the vector.